### PR TITLE
Proposal: Docker plugins MVP

### PR DESCRIPTION
**Pull request by lukemarsden**
Here is a proposal for an MVP for docker plugins that unifies all the efforts in this direction.

Note that there are now - as of Monday :) three types of prototypical docker plugins:

* Volume plugins - using HTTP as extension mechanism
* Networking plugins - using an undefined extension mechanism
* API plugins - aka "powerstrip" - using HTTP as extension mechanism

This proposal suggests that we standardize on HTTP for an initial pass on the plugins framework.
For purely local plugins using a UNIX domain socket, this is sufficiently secure.

Powerstrip has demonstrated that HTTP makes it very easy for plugin authors to get started.

I suggest we could then iterate on this choice of transport once we've got some experience prototyping these plugins.

More important is getting rough consensus and running code after agreeing the message formats that need to be sent and received between docker and plugins for volumes, networking & API extensions.

### CODE REVIEW

**Comment by shykes** [comment]
I would prefer a UI where the user doesn't need to manually specify the type of extension point. Rather, the plugin should have a way to register extension points.

So instead of `volume:foo` it should be just `foo`.

**Comment by shykes** [comment]
Also I think we should have a full-blown `plugin` subcommand eg. `docker plugin load`

**Comment by lukemarsden** [comment]
Note that if this PR were to actually get merged into the docs, it would need to be split into (at least) "using plugins" and "writing plugins" documents (for different audiences). And probably it would also make sense to split it out into a document per plugin type, at least for the "writing plugins" part.

**Comment by lukemarsden** [comment]
> I would prefer a UI where the user doesn't need to manually specify the type of extension point. Rather, the plugin should have a way to register extension points.

@shykes OK, makes sense. This would suggest that we need to implement what we're calling "plugin negotiation" where when docker first connects to the plugin, it does a sort of handshake to figure out which extension points the plugin wants to be registered for.

**Comment by lukemarsden** [comment]
> Also I think we should have a full-blown plugin subcommand eg. docker plugin load

@shykes OK, that's an easy change.

**Comment by lukemarsden** [comment]
Review comment from Adrian: perhaps better to have an explicit configuration file, so that the Docker daemon itself can drive the lifecycle rather than putting this in the hands of the user. (Order dependencies, etc.)

File: docs/sources/userguide/plugins.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+In future, we will want to extend this to notify volume extensions when containers start and stop, so that the underlying volumes mechanism can hold and release *leases* on volumes.
+
+### Volumes CLI
+
+We may want to extend the volumes CLI in due course, so you can do something like the following:
+
+```
+$ docker volume create --driver flocker \
+	--metadata size=50G,tier=ssd,replication=global nicevol
```
Comment (robhaswell) [inline docs/sources/userguide/plugins.md]: This should match the --env/--volume semantics:

`--metadata size=50G --metadata tier=ssd --metadata replication=global`

Comment (cpuguy83) [inline docs/sources/userguide/plugins.md]: Or more really the forthcoming label API

File: docs/sources/userguide/plugins.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+ Action: "create",
+ HostPath: "/path",
+ ContainerID: "abcdef123"}
+```
+
+**Response**
+
+```
+{DockerVolumesExtensionVersion: 1,
+ ModifiedHostPath: "/newpath"}
```
Comment (cpuguy83) [inline docs/sources/userguide/plugins.md]: Use the same data structure in response as the request

Comment (lukemarsden) [inline docs/sources/userguide/plugins.md]: @progrium suggests making this be the same data structure as the request.

File: docs/sources/userguide/plugins.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+**Implementation note:** Whether `docker network` calls the Docker client binary and Docker daemon, or whether the first class networks functionality is implemented by separate binary *should be transparent to the plugin*.
+
+Similar request-response pairs as for volumes can be defined for network `create`, `add-container`, etc.
+
+
+# Semantics
+
+This area needs fleshing out.
+
+* Should chaining be possible?
```
Comment (lukemarsden) [inline docs/sources/userguide/plugins.md]: @progrium suggest that this behaviour should depend on the extension point.

File: docs/sources/userguide/plugins.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+Plugin flocker:v0.1 loaded and registered with extension points:
+* volumes
+Plugin flocker activated.
+```
+
+This is really just syntactic sugar for the following:
+
+```
+$ docker run -d -e $ARGS \
+	-v /var/run/docker-plugins/flocker.sock:/var/run/plugin.sock \
```
Comment (bfirsh) [inline docs/sources/userguide/plugins.md]: Perhaps `/var/run/docker/` should be a thing? `/var/run/docker/api.sock`, `/var/run/docker/plugins/`, etc

Comment (shykes) [inline docs/sources/userguide/plugins.md]: why not use /var/lib/docker ?

File: docs/sources/userguide/plugins.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+HTTP over a local UNIX socket is used to communicate between Docker and its plugins.
+
+Plugins are distributed as containers.
+
+This is targeted at Docker 1.7.
+
+# Usage
+
+Plugins can be loaded using a special `docker plugin load` command, as follows:
```
Comment (progrium) [inline docs/sources/userguide/plugins.md]: After discussing general usefulness of hidden/system/agent containers and that they would pretty much all operate with the same semantics of plugins, maybe we can call this subcommand `agent`? To then decouple this (also so it can be a standalone PR) from plugins, we can make the plugin handshake protocol optional with a flag. Another flag would provide the Docker socket as discussed.

Comment (shykes) [inline docs/sources/userguide/plugins.md]: What is an agent?

Comment (progrium) [inline docs/sources/userguide/plugins.md]: I guess this is missing context. There is a very common pattern in deployments of "system" containers or "agents", whether custom or from a third party that do various things for the system, but is not part of the main compute payload of a host. 

Many examples are candidates for "plugins" but not all of them necessarily need to hook into Docker extension points. They just need to 1) always run on Docker boot, 2) likely have a default restart policy, 3) not show in main `docker ps`. Their management semantics match plugins but are not always plugins. For example, Docker Swarm is an agent. 

Another example of this pattern shows up in RancherOS, how they've separated system containers from user containers. Our terminology is "agent".

**Comment by bfirsh** [comment]
I wonder if these things should be called "drivers". A networking "extension" doesn't feel like it's extending the functionality of networking, it's more defining how networking works. In the same way I would install a "driver" for my network interface on my operating system – that thing wouldn't be an "extension" of the operating system, it's an essential piece of functionality.

You also hint at that when you say `docker volume create --driver flocker`.

**Comment by progrium** [comment]
It's true some things should be called drivers. The problem is, not everything is a driver. The general framework of extension points means software can be extended in more ways than what a driver does. For example, if a plugin were to extend the CLI or API, would it be a driver? However, it makes perfect sense to call a particularly type of extension point a "network driver" or "storage driver".

**Comment by lukemarsden** [comment]
# Naming

> agent vs plugin

I think the user-facing value here is very much about "plugins". The community, etc thinks we're working on plugins. If we come out with "agents" I think that will be a confusing message for people. So I vote to keep them called plugins. (People can still implement agent-like functionality with null-plugins, that's fine.)

> driver vs plugin

I agree with @progrium that not every plugin is a driver. For example, a plugin that makes the remote API only respond with GET requests isn't really "driving" anything, but could still be useful.

So I vote for continuing to use "plugin" and "extension" exchangeably. Perhaps we even want to standardize on "plugin" because it's shorter and more natural.

# Paths

@shykes are you suggesting we store `.sock` files for the plugins in `/var/lib/docker`? That's fine by me, I just want to confirm that's what you meant.

# Race conditions

We agreed on the call that from the moment a plugin is loaded, the guarantee from the docker daemon is that it won't process any API requests until the all plugins are ready (have completed handshake). When the docker daemon restarts, it will still have all plugins loaded, and will always wait until they are ready before processing any API requests (preferably, it will block those requests until ready, rather than responding with a 500).

@mrjana Does this address your concern about race conditions?

**Comment by lukemarsden** [comment]
# Work plan

@progrium to work on loading mechanism.
@cpuguy83 and @lukemarsden (that's me) to work on volumes extensions (update to latest spec).

@icecrime and @aluzzardi do you have any cycles available to work on this?

File: docs/sources/userguide/plugins.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+$ vagrant up --provider=aws
+$ vagrant ssh node1
+node1$ docker run -v /flocker/test:/data ubuntu sh -c "echo fish > /data/file.txt"
+node1$ exit
+$ vagrant ssh node2
+node2$ docker run -v /flocker/test:/data ubuntu cat /data/file.txt
+fish
+```
+
+In future, we will want to extend this to notify volume extensions when containers start and stop, so that the underlying volumes mechanism can hold and release *leases* on volumes.
```
Comment (binocarlos) [inline docs/sources/userguide/plugins.md]: If an extension has access to the docker socket - it can listen to the event stream to know if a container has stopped.  Is it better to leave this up to the extension?  The same issue would be true of networking extensions that want to release an IP.

File: docs/sources/userguide/plugins.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+
+# Semantics
+
+This area needs fleshing out.
+
+* Should chaining be possible?
+* What happens when you load several of the same type of extension?
+* What happens when a hook or pre-hook fails?
+  What about post-hooks?
```
Comment (binocarlos) [inline docs/sources/userguide/plugins.md]: Post hooks are important - IMHO they should certainly be included somehow in the workflow.  I think about it as pre-hook=setup and post-hook=react where the react step is "do something now that container is running"

File: docs/sources/userguide/plugins.md [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+  Do we need special cleanup hooks?
+
+# Possible future functionality
+
+* While waiting for the plugin to initialize, Docker should not allow any API requests to succeed.
+  This is to avoid startup race where e.g. `create` requests might not get passed through the plugins.
+* Plugins could use a protocol other than HTTP.
+* Loading a plugin marks the plugin's container as hidden from `docker ps`.
```
Comment (binocarlos) [inline docs/sources/userguide/plugins.md]: Is this an example of the more generic "agent" pattern discussed by @progrium above?

**Comment by progrium** [comment]
To comment on @binocarlos and @lukemarsden 's recent comments on agents:

I am willing to bet very strongly that `docker plugins` or whatever we call it will be used *significantly more* for the agent use case than plugins. 

This should be easy to reason given writing a plugin (by which I mean use the extension point system / plugin protocol) is more involved than an agent, and there is *lots* to do that can be done without being a plugin. How many plugins using plugin protocol can you imagine that you have a high degree of confidence in being useful and popular? Some? Obviously that's our job right now. However, whatever that number is, the number of non-plugin agents is likely *already* greater. Here are some examples that exist today:

https://github.com/gliderlabs/logspout
https://github.com/gliderlabs/registrator
https://github.com/progrium/docker-consul
https://github.com/progrium/ambassadord
https://github.com/docker/swarm/
https://github.com/coreos/etcd
https://github.com/DataDog/docker-dd-agent
https://github.com/shipyard/shipyard-agent
http://docs.aws.amazon.com/AmazonECS/latest/developerguide/ECS_agent.html
https://github.com/sequenceiq/docker-serf
https://github.com/rancherio/rancher/tree/master/agent (also rancher runs every host/system daemon in docker ... examples of what could be called "agents")
https://github.com/blalor/docker-hosts
https://github.com/antonlindstrom/docker-statsd
https://registry.hub.docker.com/u/sysdig/agent/
https://registry.hub.docker.com/u/mmcc/runscope-agent/

Plenty of projects that would normally be run as "system agents" or "system services" fall under this "agent" pattern. Think of all the one-off containers people run on their system for admin tasks ... pulling container, cleaning up images, rotating logs, etc. That's what happens when you take a 100% Docker approach. Could these all be called "plugins"? Maybe. But my point is there are lots of these ... more than there are what we are talking about here. I just want y'all to take this use case seriously.

**Comment by lukemarsden** [comment]
Hey @progrium - yes definitely. I think the case for encouraging people to write system agents is strong. And I think we agree that these can manifest as plugins with no interests (and that `/var/run/docker.sock` is always bind-mounted into plugin containers).

I propose that for a first pass it's sufficient for the docs for plugins say "system containers/agents can be implemented as plugins which register no interested subsystems at handshake time".

**Comment by progrium** [comment]
Sounds Good Enough (tm)

File: docs/sources/userguide/plugins.md (hunk 6 above)
Comment (lukemarsden) [inline docs/sources/userguide/plugins.md]: Yes I think that's OK for the plugin to implement in its own way.

File: docs/sources/userguide/plugins.md (hunk 7 above)
Comment (lukemarsden) [inline docs/sources/userguide/plugins.md]: This is part of the API extension discussion then. When we get to that bit, we can start with what we learned from powerstrip, write down the desired formal semantics and iterate on that.

File: docs/sources/userguide/plugins.md (hunk 8 above)
Comment (lukemarsden) [inline docs/sources/userguide/plugins.md]: Yes, I think we agreed to go ahead and start with this (it's up to @progrium how he implements the loader :smile:).

**Comment by mrjana** [comment]
@lukemarsden Regarding the race condition aspect, imagine a plugin hooking to /container/create and /container/start(very similar to powerstrip usecase) but the remote API for /container/create was issued by a client before any plugin was loaded and so docker core handled it and returns and then the plugins got loaded and after that the the client issued a /container/start which the docker daemon will dutifully invoke the plugin. Now in this situation if a plugin was expecting to see both /container/create and /container/start in a container life cycle, but instead only see /container/start then it's going to need some complicated logic in the plugin to handle this. Makes sense or am I totally way off?

**Comment by lukemarsden** [comment]
@mrjana this is true, but I don't think there's much we can do about it. It's always possible that we're between two semantically dependent API calls so there's never a "good" time to load a plugin.

I suggest that we solve this issue by encouraging in the documentation that users load plugins "when the docker daemon is quiet, for example during a provisioning phase". And that users "take the docker daemon out of service before upgrading plugins".

I propose that we defer thinking too hard about this until there's a bit more progress on this topic. Unless you have an alternative suggestion?

**Comment by shykes** [comment]
Just to be clear I am commenting here on a best effort basis, I'm mostly ignoring this to let you guys prepare a proposal. In other words don't imply agreement by default ;)

**Comment by mrjana** [comment]
@lukemarsden I am fine with "making progress" part, but documenting users to load plugins when the docker daemon is quiet may not always be practical especially when the whole system becomes highly dynamic with all the APIs which are driven by a bunch of orchestration agents and other intelligent pieces of software creating a complete Machine2Machine interaction.

One alternative I would suggest to add a docker UI enhancement to accept a plugin configuration file which will be read by the daemon during init time (at which time it will not accept any APIs) and then load all the plugins. Atleast that provides a way to guarantee avoiding interleaved plugin loads between semantically dependent API invocations

**Comment by shykes** [comment]
I don't think we will need to tell users to load plugins when the daemon is
quiet. If your plugin needs to hook "create" to work properly on "start",
it's your plugin's responsibility to check for that (for example by looking
for a special annotation key it left for itself at the previous stage). If
a plugin doesn't check for that and introduces bugs as a result, it's a bug
in the plugin.

I think that should take care of the "plugin loaded half-way through the
sequence of calls" problem.

**Comment by lukemarsden** [comment]
@mrjana I worry that using a config file doesn't help either. There's nothing stopping docker from being restarting docker during a create => start window anyway. I think @shykes is right: it is on the plugin author to store state needed to guarantee its own internal consistency even when it might not receive all events.

**Comment by lukemarsden** [comment]
> Just to be clear I am commenting here on a best effort basis

@shykes understood, no problem.

File: docs/sources/userguide/plugins.md [hunk 9]
```diff
@@ -0,0 +1,157 @@
+Docker Plugins MVP
+==================
+
+Plugins provide a mechanism for hooking into the Docker engine to extend its behavior.
+
+HTTP over a local UNIX socket is used to communicate between Docker and its plugins.
+
+Plugins are distributed as containers.
```
Comment (dave-tucker) [inline docs/sources/userguide/plugins.md]: I think packaging plugins as containers is great - it saves a lot of work from the plugin authors side.
However, The question I have is about *how* they are distributed. Today the Docker registry doesn't contain any data about metadata about a container and what permissions it requires. This metadata is important from a users standpoint (a parallel would be the Google Play Store for Android) so they can make a decision of whether or not to use a plugin.

As @progrium mentions below, most agents will require `--privileged` or perhaps a combination of `--cap-add` and `--cap-drop`.

Perhaps a plugin should be a container + a `docker-compose.yml`

File: docs/sources/userguide/plugins.md (hunk 4 above)
Comment (dave-tucker) [inline docs/sources/userguide/plugins.md]: i've always thought that sockets and pid files live in `/var/run` or `/usr/var/run`
at least, that's a pattern I've come to expect when using linux

File: docs/sources/userguide/plugins.md [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+## Network
+
+Network extensions exist to implement the following UX:
+
+```
+$ docker network create greatnet
+$ docker run --net=greatnet --ip=1.2.3.4 dockerfile/postgresql
+```
+
+(and of course: `$ docker run --net=greatnet --ip=1.2.3.4 -v //nicevol:/data dockerfile/postgresql` should also be possible, when both networking and storage plugins are loaded)
```
Comment (dave-tucker) [inline docs/sources/userguide/plugins.md]: i'm wondering whether it would be useful for a plugin to register a custom uri scheme.
So you could chain plugins like so...

``` bash
docker run --net=weave://network1 --net=other://network12 -v flocker://volume-foo:/usr/foo -v other://volume-bar:/usr/bar
```

File: docs/sources/userguide/plugins.md (hunk 4 above)
Comment (cpuguy83) [inline docs/sources/userguide/plugins.md]: On the host side, I think inside `/var/run/docker/plugins/<id>` would be the place to put that.
In the container, /run or /var/run seems like the right thing.

File: docs/sources/userguide/plugins.md (hunk 9 above)
Comment (cpuguy83) [inline docs/sources/userguide/plugins.md]: I think this is something we can work towards... basically having the image declare the permissions it needs to be able to run and error out if those perms weren't granted (ie by `docker plugin load --privileged`).

But I don't see this as a hard requirement for plugins.

Comment (lukemarsden) [inline docs/sources/userguide/plugins.md]: Agree. Caveat here is that if we bind-mount the docker socket, then any plugin can elevate its own privileges. This needs more thought but not yet for an extensions MVP.

Comment (shykes) [inline docs/sources/userguide/plugins.md]: We should definitely not bind-mount the docker socket into the plugins.
Once people start writing plugins which assume this, it will be impossible
to change the design without breaking them.

Comment (cpuguy83) [inline docs/sources/userguide/plugins.md]: I agree with not automatically doing it for all plugins, it would be nice to request one (not using `-v /var/run/docker.sock:/var/run/docker.sock`) from the UI.

Comment (lukemarsden) [inline docs/sources/userguide/plugins.md]: Hmm. The reason we wanted the Docker socket was so that e.g. extension payloads (the messages docker sends to the extensions) could refer to e.g. a container ID and the extension would be free to look up the data about the container. `powerstrip-weave` also had a use-case where it was necessary to inspect the image JSON for a container as it was being started.

A more sophisticated approach than bind-mounting the docker socket would be to define a set of interfaces with different sets of permissions. So for this use case we could have a "getContainerJSON" permission. Then we can fill this catalogue of interfaces as we come across them.

Or a middle ground could be to:
a) define permissions, make it explicit that permissions get granted somehow, and
b) define a "can have read/write access to the Docker API" permission.

I'm really keen to keep this MVP so that we can land it in Docker 1.7. I suggest we approach it therefore with the above middle-ground: that way when users load extensions, they have to explicitly give the plugin permission.

Note that plugins will often want to have (effectively, via `--privileged`) root on the host anyway to do system level things.

@progrium @shykes what do you think?

Comment (shykes) [inline docs/sources/userguide/plugins.md]: I think it's too early to invent a whole new permission system, and a ui
for it, on a short deadline. I don't think we need it for MVP.

We already have an existing permission system: the options to "docker run"
and "docker start". Privileged, net=none, capabilities etc. Since a plugin
is a container, it stands to reason that plugin permissions should simply
be container permissions.

File: docs/sources/userguide/plugins.md (hunk 10 above)
Comment (lukemarsden) [inline docs/sources/userguide/plugins.md]: +1 for the URI scheme in a later revision.

I've been toying with `/flocker/volume-foo`, `//volume-foo`, `flocker://volume-foo` and I like `flocker://volume-foo` the best. The reason I hadn't proposed that yet though is that it would require a change to the parsing logic because colons already have a meaning in volume specs. `/flocker/volume-foo` certainly isn't final, but it will get us over the hurdle of getting an extensions MVP out.

File: docs/sources/userguide/plugins.md (hunk 9 above)
Comment (progrium) [inline docs/sources/userguide/plugins.md]: Also against permissions. But I can't stress enough how important the Docker socket is here. I don't think it should be included by default, but I do really really really want a convenience flag for it. I don't care if it is someday deprecated or removed, assuming it is replaced by something else that works (limited permissions, etc)

Comment (lukemarsden) [inline docs/sources/userguide/plugins.md]: I agree with this @shykes. Bind-mounting (which is how we'd propose to expose the Docker API) is already arguably part of the docker permissions interface (let the container mess with this subtree on the host filesystem). Does that mean that the plugin loader interface should have `-v`?

To be clear, the choice is:
* generic `-v` support for plugin loader.
* some sort of specific `--allow-docker-api` flag to the plugin loader which special-cases bind-mounting in the docker sock.

I think @progrium preferred the latter. I'm cool with either. If you strongly prefer the former @shykes I'm fine with that.

One data point: the prototypical containerized `flocker-zfs-agent` I'm working on currently has 5 bind-mounts. BTW, I was imagining that the flocker extension would start the `flocker-zfs-agent` with its requisite bind-mounts via the docker api when it started... but we really do need the extensions to be able to "do stuff". I don't think we have time to define an interface with granular permissions. And in fact arguably the interface with granular permissions should be a later version of the Docker API!

Comment (lukemarsden) [inline docs/sources/userguide/plugins.md]: As a point of protocol, I think @progrium is working on the loader spec this/next week. It may be more productive to let him finish that and then do a review pass over it than to pre-empt his proposed design, which I'm sure will handle the use cases we have in mind. ;)

File: docs/sources/userguide/plugins.md (hunk 10 above)
Comment (cpuguy83) [inline docs/sources/userguide/plugins.md]: So we've talked about this in the past and it the response was basically this was overloading the `-v`... that said I think it totally makes sense.

File: docs/sources/userguide/plugins.md (hunk 9 above)
Comment (binocarlos) [inline docs/sources/userguide/plugins.md]: +1 for having **some** way for a plugin to speak to the main docker server - it is vital for various reasons - I'm not overly concerned with how though (`-v` or `--allow-docker-api`)

File: docs/sources/userguide/plugins.md [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+* Should chaining be possible?
+* What happens when you load several of the same type of extension?
+* What happens when a hook or pre-hook fails?
+  What about post-hooks?
+  Do we need special cleanup hooks?
+
+# Possible future functionality
+
+* While waiting for the plugin to initialize, Docker should not allow any API requests to succeed.
```
Comment (squaremo) [inline docs/sources/userguide/plugins.md]: Won't this prevent plugins from doing any work through the Docker API? For instance, when a weave plugin was loaded, we'd want it to start a weave router container, maybe a weaveDNS container, and so on.

File: docs/sources/userguide/plugins.md [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+Extensions to the Docker remote API (extension type `api`) can be achieved via a [Powerstrip-like protocol](https://github.com/clusterhq/powerstrip#pre-hook-adapter-endpoints-receive-posts-like-this).
+The corresponding limitations will initially apply to the types of requests which can be processed, but this is acceptable for an MVP.
+
+## Network
+
+Network extensions exist to implement the following UX:
+
+```
+$ docker network create greatnet
```
Comment (squaremo) [inline docs/sources/userguide/plugins.md]: Is `greatnet` the name of the plugin, or the name of the network?

File: docs/sources/userguide/plugins.md [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+```
+{HostPath: "/path",
+ ContainerID: "abcdef123"}
+```
+
+**Response**
+
+```
+{HostPath: "/newpath",
+ ContainerID: "abcdef123"}
```
Comment (squaremo) [inline docs/sources/userguide/plugins.md]: Why is the `ContainerID` here -- can the plugin change it? What would that mean?

File: docs/sources/userguide/plugins.md (hunk 11 above)
Comment (binocarlos) [inline docs/sources/userguide/plugins.md]: This is a good point - a plugin might itself need to start other containers.  Perhaps we can say "While waiting for the plugin to initialize, Docker should not allow any API requests **that are not from plugins** to succeed".  This is also another good use-case for plugins in general being able to talk to the docker deamon

File: docs/sources/userguide/plugins.md [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+Plugins are distributed as containers.
+
+This is targeted at Docker 1.7.
+
+# Usage
+
+Plugins can be loaded using a special `docker plugin load` command, as follows:
+
+```
+$ docker plugin load -e $ARGS clusterhq/flocker-plugin $@
```
Comment (errordeveloper) [inline docs/sources/userguide/plugins.md]: why should this have `-e $ARGS` and `$@`? I suppose these are placeholders for the launch configuration... The line is a little confusing, I think if it just said `docker plugin load clusterhq/flocker-plugin`, it'd totally fine. You can hint about configuration though, which if you want can be through arguments. Although, keeping configuration out of the scope for now would be good, you could suggest that configuration would be part of the image if you like.

File: docs/sources/userguide/plugins.md (hunk 10 above)
Comment (errordeveloper) [inline docs/sources/userguide/plugins.md]: This line also introduces the `--ip` argument, which rises another question - how would user know that this argument is handled by the extension? One way of solving this problem is to add extend `--net="..."` syntax, but that is gonna get hairy soon. Perhaps plugins could introduce a namespaced argument, e.g. `--awesome-ip=` in this specific case would work a bit better.

Comment (errordeveloper) [inline docs/sources/userguide/plugins.md]: I do like the URI approach in ways, but extending syntax of `--net` is not a great idea, I would rather deprecate that flag and introduce something like this:
```
docker run --weave-net=network1 --other-net=network12 --other-param=foo
```
Similarly with volumes, I am thinking that adding `--flocker-volume` would appear much clearer to the user.

File: docs/sources/userguide/plugins.md (hunk 12 above)
Comment (lukemarsden) [inline docs/sources/userguide/plugins.md]: The name of the network, I guess. This is where my view of what @shykes has in mind for the networking UX gets pretty hazy.

File: docs/sources/userguide/plugins.md (hunk 13 above)
Comment (lukemarsden) [inline docs/sources/userguide/plugins.md]: It's there so that the plugin can look up (or record) the container info if it needs to.

File: docs/sources/userguide/plugins.md (hunk 14 above)
Comment (lukemarsden) [inline docs/sources/userguide/plugins.md]: The `-e $ARGS` was for environment variables to pass to the plugin, and `$@` just to say that you can pass some arguments. Should we just pick one of these? Or have neither?

Having config as part of the plugin image doesn't make sense because it's config provided by the cluster admin not the plugin author.

Comment (errordeveloper) [inline docs/sources/userguide/plugins.md]: > Should we just pick one of these? Or have neither?

I think having args at the end will be fine for the purpose of this document.

> Having config as part of the plugin image doesn't make sense...

Well, the idea was to reduce the scope of this proposal, there is all sorts of things one can dream up about config, but it's the kind of thing that is always easy to over-engineer. For example, this can easily get everyone into bikeshedding a distributed config store or syntax format, and that's best to avoid if progress is to be made on introducing the primitives, into which this is already getting deep enough.

File: docs/sources/reference/plugin-mvp.md [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+ Name: "flocker",
+ Author: "Luke Marsden <luke@clusterhq.com>",
+ Org: "ClusterHQ, Inc.",
+ Website: "https://clusterhq.com/",
+}
+```
+
+`PluginName` is a human readable short string identifying the plugin.
+
+`InterestedIn` is a list of extension points.
```
Comment (progrium) [inline docs/sources/reference/plugin-mvp.md]: Perhaps "implements"?

File: docs/sources/userguide/plugins.md (hunk 9 above)
Comment (icecrime) [inline docs/sources/userguide/plugins.md]: My take on these:

- I'd rather *not* expose `docker run` arguments to `docker plugin` for two reasons
   - As soon as we start adding some of them, we'll end up adding all of them.
   - I like that we have a dedicated `docker plugin` top-level command that hides the fact that it's actually pulling an image and running a container: I believe makes perfect sense for an MVP, but I don't think we'd like to lock ourselves in that as the only possible execution model.
- Mounting the daemon socket should not be the default, although I agree it should be made possible through some command-line flag that makes it obvious you're doing something "privileged", such as `--allow-full-daemon-access`.

File: docs/sources/userguide/plugins.md (hunk 13 above)
Comment (squaremo) [inline docs/sources/userguide/plugins.md]: I mean in the reply

File: docs/sources/reference/plugin-mvp.md (hunk 15 above)
Comment (lukemarsden) [inline docs/sources/reference/plugin-mvp.md]: +1

**Comment by lukemarsden** [comment]
When chatting with @binocarlos and @squaremo we discussed the need to perhaps introduce a concept of "read" or "write" access to a subsystem. This might help with chaining.

**Comment by progrium** [comment]
How would you model "read" or "write" access with a native language API?

**Comment by cpuguy83** [comment]
I think if you are registering a plugin you should make sure it only does what you want it to.
If you only give a plugin read access and it was expecting to be able to write, it will break.

So basically, the plugin has to be written to only care about reads... in which case is there any sense in enforcing this?

I suppose a read-only plugin is one that is just a fire-and-forget... ie docker doesn't wait for a response from it.

**Comment by progrium** [comment]
Again, I believe this should be address in the design of the extension point interface by the subsystem author, not the responsibility of this protocol / system.

**Comment by squaremo** [comment]
> we discussed the need to perhaps introduce a concept of "read" or "write" access to a subsystem

This was one way of accounting for composing extensions: if more than one plugin registers for an extension point, one can have write and the others only read, and docker breaks when loading the overreaching plugin if you get it wrong.

Another way to account for composition: don't have it. Only one plugin can be loaded for each extension point, and docker breaks when loading the extraneous plugin if you get it wrong.

Another way: Specify which extension is being called, e.g., `docker run --net=weave/args ...`, and docker (run) breaks if you name a plugin that's not loaded.

Something to keep in mind is how surprising these things might be to users. Because of the "register interest" handshaking, it won't necessarily be obvious to a user that e.g., the flocker plugin is a "volume" plugin and they can have only one of those (or only one "volume" write plugin).

File: docs/sources/reference/plugin-mvp.md [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+Extension points currently supported are:
+
+* `volume` - called when a volume is bind-mounted into a container at create or start time
```
Comment (duglin) [inline docs/sources/reference/plugin-mvp.md]: I assume we can replace "when" with "after" ?  just to be clear about when it is called.

File: docs/sources/reference/plugin-mvp.md [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+**Response**
+
+```
+{HostPath: "/newpath",
+ ContainerID: "abcdef123"}
+```
+
+# Caveats
+
+NB: plugin loading order is undefined, so plugins should not depend on eachother.
```
Comment (duglin) [inline docs/sources/reference/plugin-mvp.md]: Can people assume they are called in the order in which they are registered?

File: docs/sources/userguide/plugins.md (hunk 9 above)
Comment (duglin) [inline docs/sources/userguide/plugins.md]: While I like the idea of a container for plugins on the daemon-side, I'm wondering about CLI plugins. I don't think they can use the same mechanism. I wonder if it makes sense to have cli-plugins be executables that use the same API defined in there?  Meaning, via some cli-specific registration mechanism (perhaps they're listed in the cli config file), the cli can start the exe's and then talk to them via HTTP using the same protocol defined here.  So while we can't get consistency around distribution, we can be consistent on the API/protocol.

File: docs/sources/userguide/plugins.md [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+The simplest of plugin types, `volume` provides a single request-response type on `POST /v1/volumes` (`POST ~= create`):
+
+**Request**
+
+```
+{HostPath: "/path",
+ ContainerID: "abcdef123"}
+```
+
+**Response**
```
Comment (duglin) [inline docs/sources/userguide/plugins.md]: Can we get some text around what the values in the response mean?  It would seem to imply that the plugin can change the hostpath and container id - but that's doesn't seem right.  Also, above I was assuming this is called after the volume is mounted, but this response (if these values are modifiable by the plug-in) would seem to imply its called before its mounted.  For each plug-point we need to be clear about the "before" vs "after" nature of each.

Comment (duglin) [inline docs/sources/userguide/plugins.md]: We may want to consider responses that stop the current action.  For example, does the mount stop if the plug-in fails?  Do we allow the plug-in to stop the action at all?  What happens if it fails? Do we roll things back?
My initial guess is that a failed plug-in should stop the flow - otherwise users could get unexpected results - or worse, very bad results if one of the plug-ins is an auth-type check.  And therefore, if possible, I think things should rollback.
